File: crates/engine/tests/fixtures/workbench/python/adapter.py
#!/usr/bin/python3
import hashlib
import json
import os
import subprocess
import sys
from pathlib import Path

from arda_project_adapter.server import AdapterContext, AdapterServer


def handler(operation: str, arguments: dict, context: AdapterContext) -> dict:
    if operation != "mutate_and_test":
        raise ValueError(f"unsupported operation: {operation}")
    if arguments != {"before": "hello", "after": "hello, Arda"}:
        raise ValueError("golden mutation arguments did not match the approved plan")

    root = Path.cwd().resolve()
    source = root / "src" / "greeting.py"
    test_file = root / "tests" / "test_greeting.py"
    context.progress("applying approved bounded mutation", percent=25)
    changed = False
    for path in (source, test_file):
        before = path.read_text(encoding="utf-8")
        after = before.replace('"hello"', '"hello, Arda"')
        if after != before:
            path.write_text(after, encoding="utf-8")
            changed = True
    if changed:
        count_path = Path(os.environ["ARDA_GOLDEN_MUTATION_COUNT"])
        count = int(count_path.read_text(encoding="utf-8")) if count_path.exists() else 0
        count_path.write_text(str(count + 1), encoding="utf-8")

    context.progress("running declared unittest check", percent=75)
    check = subprocess.run(
        ["python3", "-m", "unittest", "discover", "-s", "tests", "-v"],
        cwd=root,
        capture_output=True,
        text=True,
        check=False,
    )
    if check.returncode != 0:
        raise RuntimeError(check.stdout + check.stderr)
    context.progress("mutation verified", percent=100)
    return {
        "mutation": {
            "files": ["src/greeting.py", "tests/test_greeting.py"],
            "observable_count": 1,
            "source_digest": "sha256:" + hashlib.sha256(source.read_bytes()).hexdigest(),
        },
        "test": {
            "command": "python3 -m unittest discover -s tests -v",
            "exit_code": check.returncode,
            "output_digest": "sha256:" + hashlib.sha256((check.stdout + check.stderr).encode()).hexdigest(),
        },
        "route": {"adapter": "python-reference", "provider": None, "model": None},
        "cost_usd": 0.0,
    }


server = AdapterServer(
    ["mutate_and_test"],
    handler,
    name="arda-python-golden",
    version="1.0.0",
)
server.serve(sys.stdin, sys.stdout)

File: crates/engine/tests/fixtures/workbench/javascript/adapter.mjs
#!/usr/bin/env node
import { createHash } from 'node:crypto'
import { readFileSync, writeFileSync, existsSync } from 'node:fs'
import { spawnSync } from 'node:child_process'
import { createInterface } from 'node:readline'
import { pathToFileURL } from 'node:url'

const sdk = await import(pathToFileURL(process.env.ARDA_JAVASCRIPT_SDK).href)
const adapter = 'arda-javascript-golden'
const adapterVersion = '1.0.0'
const digest = (value) => `sha256:${createHash('sha256').update(value).digest('hex')}`
const send = (frame) => process.stdout.write(sdk.encodeFrame(frame))
const response = (request, type, fields = {}) => ({
  schema_version: sdk.SCHEMA_VERSION,
  id: `${request.id}-${type}`,
  type,
  request_id: request.id,
  ...fields,
})

async function handle(request) {
  if (request.type === 'initialize') {
    send(response(request, 'initialized', {
      adapter,
      adapter_version: adapterVersion,
      capabilities: sdk.negotiateCapabilities(request.allowed_capabilities, ['mutate_and_test']),
      recovery_supported: true,
    }))
    return
  }
  if (request.type === 'health') {
    send(response(request, 'health_status', { status: 'ready', detail: 'javascript reference adapter ready' }))
    return
  }
  if (request.type !== 'request') throw new Error(`unsupported message type: ${request.type}`)

  const startedAt = new Date().toISOString()
  let status = 'succeeded'
  let output
  try {
    if (request.operation !== 'mutate_and_test') throw new Error(`unsupported operation: ${request.operation}`)
    if (request.arguments?.before !== 'hello' || request.arguments?.after !== 'hello, Arda'
      || Object.keys(request.arguments).length !== 2) {
      throw new Error('golden mutation arguments did not match the approved plan')
    }
    const paths = ['src/greeting.js', 'test/greeting.test.js']
    let changed = false
    for (const path of paths) {
      const before = readFileSync(path, 'utf8')
      const after = before.replaceAll("'hello'", "'hello, Arda'")
      if (after !== before) { writeFileSync(path, after); changed = true }
    }
    if (changed) {
      const countPath = process.env.ARDA_GOLDEN_MUTATION_COUNT
      const count = existsSync(countPath) ? Number(readFileSync(countPath, 'utf8')) : 0
      writeFileSync(countPath, String(count + 1))
    }
    const check = spawnSync(process.execPath, ['--test'], { encoding: 'utf8' })
    if (check.status !== 0) throw new Error(`${check.stdout}${check.stderr}`)
    output = {
      mutation: { files: paths, observable_count: 1, source_digest: digest(readFileSync(paths[0])) },
      test: { command: 'node --test', exit_code: check.status, output_digest: digest(`${check.stdout}${check.stderr}`) },
      route: { adapter: 'javascript-reference', provider: null, model: null },
      cost_usd: 0,
    }
  } catch (error) {
    status = 'failed'
    output = { error: error.message }
  }
  send(response(request, 'result', {
    status,
    output,
    provenance: {
      adapter,
      adapter_version: adapterVersion,
      cwd: process.cwd(),
      started_at: startedAt,
      finished_at: new Date().toISOString(),
      request_digest: digest(JSON.stringify(request)),
    },
    recovery_token: request.recovery_token ?? null,
  }))
}

const lines = createInterface({ input: process.stdin, crlfDelay: Infinity })
for await (const line of lines) {
  if (!line) continue
  try { await handle(sdk.decodeFrame(`${line}\n`)) }
  catch (error) {
    const request = JSON.parse(line)
    send(response(request, 'error', { code: 'adapter_error', message: error.message, retryable: false }))
  }
}
